File: android/src/main/java/com/asterinet/react/bgactions/RNBackgroundActionsTask.java
package com.asterinet.react.bgactions;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ServiceInfo;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.core.app.ServiceCompat;

import com.facebook.react.HeadlessJsTaskService;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.jstasks.HeadlessJsTaskConfig;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.ArrayList;

final public class RNBackgroundActionsTask extends HeadlessJsTaskService {

    public static final int SERVICE_NOTIFICATION_ID = 92901;
    private static final String CHANNEL_ID = "RN_BACKGROUND_ACTIONS_CHANNEL";
    private @Nullable static BackgroundTaskOptions currentBgOptions;

    @SuppressLint("UnspecifiedImmutableFlag")
    @NonNull
    public static Notification buildNotification(@NonNull Context context, @NonNull final BackgroundTaskOptions bgOptions) {
        currentBgOptions = bgOptions;

        // Get info
        final String taskTitle = bgOptions.getTaskTitle();
        final String taskDesc = bgOptions.getTaskDesc();
        final int iconInt = bgOptions.getIconInt();
        final int color = bgOptions.getColor();
        final String linkingURI = bgOptions.getLinkingURI();
        Intent notificationIntent;
        if (linkingURI != null) {
            notificationIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(linkingURI));
        } else {
            //as RN works on single activity architecture - we don't need to find current activity on behalf of react context
            notificationIntent = new Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LAUNCHER);
        }
        final PendingIntent contentIntent;
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            contentIntent = PendingIntent.getActivity(context,0, notificationIntent, PendingIntent.FLAG_MUTABLE | PendingIntent.FLAG_ALLOW_UNSAFE_IMPLICIT_INTENT);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            contentIntent = PendingIntent.getActivity(context, 0, notificationIntent, PendingIntent.FLAG_MUTABLE);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            contentIntent = PendingIntent.getActivity(context, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
        } else {
            contentIntent = PendingIntent.getActivity(context, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);
        }
        final NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID)
                .setContentTitle(taskTitle)
                .setContentText(taskDesc)
                .setSmallIcon(iconInt)
                .setContentIntent(contentIntent)
                .setOngoing(bgOptions.isOngoing())
                .setAutoCancel(bgOptions.isAutoCancel())
                .setPriority(bgOptions.getPriority())
                .setColor(color);

        final Bundle progressBarBundle = bgOptions.getProgressBar();
        if (progressBarBundle != null) {
            final int progressMax = (int) Math.floor(progressBarBundle.getDouble("max"));
            final int progressCurrent = (int) Math.floor(progressBarBundle.getDouble("value"));
            final boolean progressIndeterminate = progressBarBundle.getBoolean("indeterminate");
            builder.setProgress(progressMax, progressCurrent, progressIndeterminate);
        }
        return builder.build();
    }


    @Nullable
    private ReactContext getMyReactContext() {
        ReactNativeHost reactNativeHost = getReactNativeHost();
        ReactInstanceManager reactInstanceManager = reactNativeHost.getReactInstanceManager();
        return reactInstanceManager.getCurrentReactContext();
    }

    @Override
    protected @Nullable
    HeadlessJsTaskConfig getTaskConfig(Intent intent) {
        final Bundle extras = intent.getExtras();
        if (extras != null) {
            return new HeadlessJsTaskConfig(extras.getString("taskName"), Arguments.fromBundle(extras), 0, true);
        }
        return null;
    }

    @SuppressLint("ForegroundServiceType")
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        final Bundle extras = intent.getExtras();
        if (extras == null) {
            throw new IllegalArgumentException("Extras cannot be null");
        }
        final BackgroundTaskOptions bgOptions = new BackgroundTaskOptions(extras);
        createNotificationChannel(bgOptions.getTaskTitle(), bgOptions.getTaskDesc()); // Necessary creating channel for API 26+
        // Create the notification
        final Notification notification = buildNotification(this, bgOptions);

        if (bgOptions.isOngoing()) {
            notification.flags |= Notification.FLAG_NO_CLEAR | Notification.FLAG_ONGOING_EVENT;
        }

        int type = this.getServiceType(bgOptions.getServiceTypes());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(SERVICE_NOTIFICATION_ID, notification, type);
        } else {
            startForeground(SERVICE_NOTIFICATION_ID, notification);
        }

        return super.onStartCommand(intent, flags, startId);
    }

    private void createNotificationChannel(@NonNull final String taskTitle, @NonNull final String taskDesc) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            final int importance = NotificationManager.IMPORTANCE_DEFAULT;
            final NotificationChannel channel = new NotificationChannel(CHANNEL_ID, taskTitle, importance);
            channel.setDescription(taskDesc);
            final NotificationManager notificationManager = getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        super.onTaskRemoved(rootIntent);

        try {
            WritableMap params = Arguments.createMap();
            getMyReactContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                    .emit("appKilled", params);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (currentBgOptions != null && currentBgOptions.isStopOnTerminate()) {
            stopSelf();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        try {
            WritableMap params = Arguments.createMap();
            getMyReactContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                    .emit("appKilled", params);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (currentBgOptions != null && currentBgOptions.isStopOnTerminate()) {
            stopSelf();
        }
    }

    private int getServiceType(ArrayList<String> serviceTypeArray) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
            return 0;
        }

        int type = 0;

        for (String serviceType : serviceTypeArray) {
            switch (serviceType) {
                case "camera":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_CAMERA;
                    break;
                case "connectedDevice":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_CONNECTED_DEVICE;
                    break;
                case "dataSync":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_DATA_SYNC;
                    break;
                case "health":
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                        type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_HEALTH;
                    }
                    break;
                case "location":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_LOCATION;
                    break;
                case "mediaPlayback":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK;
                    break;
                case "mediaProjection":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION;
                    break;
                case "microphone":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_MICROPHONE;
                    break;
                case "phoneCall":
                    type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_PHONE_CALL;
                    break;
                case "remoteMessaging":
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                        type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_REMOTE_MESSAGING;
                    }
                    break;
                case "shortService":
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                        type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_SHORT_SERVICE;
                    }
                    break;
                case "specialUse":
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                        type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE;
                    }
                    break;
                case "systemExempted":
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                        type |= ServiceInfo.FOREGROUND_SERVICE_TYPE_SYSTEM_EXEMPTED;
                    }
                    break;
            }
        }

        return type;
    }
}

File: android/src/main/java/com/backgroundactions/RNBackgroundActionsTask.kt
package com.backgroundactions

import android.annotation.SuppressLint
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.net.Uri
import android.os.Build
import android.os.Bundle
import androidx.core.app.NotificationCompat
import com.facebook.react.HeadlessJsTaskService
import com.facebook.react.ReactInstanceManager
import com.facebook.react.ReactNativeHost
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.jstasks.HeadlessJsTaskConfig
import com.facebook.react.modules.core.DeviceEventManagerModule

class RNBackgroundActionsTask : HeadlessJsTaskService() {

    override fun getTaskConfig(intent: Intent): HeadlessJsTaskConfig? {
        return intent.extras?.let { extras ->
            HeadlessJsTaskConfig(
                extras.getString("taskName"),
                Arguments.fromBundle(extras),
                0,
                true
            )
        }
    }

    @SuppressLint("ForegroundServiceType")
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val extras = intent.extras ?: throw IllegalArgumentException("Extras cannot be null")
        val bgOptions = BackgroundTaskOptions(extras)

        // Create notification channel for API 26+
        createNotificationChannel(bgOptions.taskTitle, bgOptions.taskDesc)

        // Create the notification
        val notification = buildNotification(this, bgOptions)

        if (bgOptions.isOngoing) {
            notification.flags = notification.flags or (Notification.FLAG_NO_CLEAR or Notification.FLAG_ONGOING_EVENT)
        }

        val type = getServiceType(bgOptions.serviceTypes ?: ArrayList())

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(SERVICE_NOTIFICATION_ID, notification, type)
        } else {
            startForeground(SERVICE_NOTIFICATION_ID, notification)
        }

        return super.onStartCommand(intent, flags, startId)
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        super.onTaskRemoved(rootIntent)
        handleTaskTermination()
    }

    override fun onDestroy() {
        super.onDestroy()
        handleTaskTermination()
    }

    private fun handleTaskTermination() {
        try {
            val params = Arguments.createMap()
            getReactContext()?.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                ?.emit("appKilled", params)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        if (currentBgOptions?.isStopOnTerminate == true) {
            stopSelf()
        }
    }

    private fun createNotificationChannel(taskTitle: String, taskDesc: String) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(CHANNEL_ID, taskTitle, importance).apply {
                description = taskDesc
            }
            getSystemService(NotificationManager::class.java)?.createNotificationChannel(channel)
        }
    }

    private fun getServiceType(serviceTypeArray: ArrayList<String>): Int {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) return 0

        return serviceTypeArray.fold(0) { acc, serviceType ->
            acc or when (serviceType) {
                "camera" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_CAMERA
                "connectedDevice" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_CONNECTED_DEVICE
                "dataSync" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_DATA_SYNC
                "health" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_HEALTH
                } else 0
                "location" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_LOCATION
                "mediaPlayback" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PLAYBACK
                "mediaProjection" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION
                "microphone" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_MICROPHONE
                "phoneCall" -> ServiceInfo.FOREGROUND_SERVICE_TYPE_PHONE_CALL
                "remoteMessaging" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_REMOTE_MESSAGING
                } else 0
                "shortService" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_SHORT_SERVICE
                } else 0
                "specialUse" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE
                } else 0
                "systemExempted" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_SYSTEM_EXEMPTED
                } else 0
                else -> 0
            }
        }
    }

    companion object {
        const val SERVICE_NOTIFICATION_ID = 92901
        private const val CHANNEL_ID = "RN_BACKGROUND_ACTIONS_CHANNEL"
        private var currentBgOptions: BackgroundTaskOptions? = null

        @SuppressLint("UnspecifiedImmutableFlag")
        fun buildNotification(context: Context, bgOptions: BackgroundTaskOptions): Notification {
            currentBgOptions = bgOptions

            val taskTitle = bgOptions.taskTitle
            val taskDesc = bgOptions.taskDesc
            val iconInt = bgOptions.iconInt
            val color = bgOptions.color
            val linkingURI = bgOptions.linkingURI

            val notificationIntent = if (linkingURI != null) {
                Intent(Intent.ACTION_VIEW, Uri.parse(linkingURI))
            } else {
                Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LAUNCHER)
            }

            val contentIntent = when {
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE -> {
                    PendingIntent.getActivity(
                        context, 0, notificationIntent,
                        PendingIntent.FLAG_MUTABLE or PendingIntent.FLAG_ALLOW_UNSAFE_IMPLICIT_INTENT
                    )
                }
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.S -> {
                    PendingIntent.getActivity(context, 0, notificationIntent, PendingIntent.FLAG_IMMUTABLE)
                }
               /* Build.VERSION.SDK_INT >= Build.VERSION_CODES.M -> {
                    PendingIntent.getActivity(
                        context, 0, notificationIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
                    )
                } */
                else -> {
                    PendingIntent.getActivity(context, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT)
                }
            }

            val builder = NotificationCompat.Builder(context, CHANNEL_ID).apply {
                setContentTitle(taskTitle)
                setContentText(taskDesc)
                setSmallIcon(iconInt)
                setContentIntent(contentIntent)
                setOngoing(bgOptions.isOngoing)
                setAutoCancel(bgOptions.isAutoCancel)
                setPriority(bgOptions.priority)
                setColor(color)

                bgOptions.progressBar?.let { progressBarBundle ->
                    val progressMax = progressBarBundle.getDouble("max").toInt()
                    val progressCurrent = progressBarBundle.getDouble("value").toInt()
                    val progressIndeterminate = progressBarBundle.getBoolean("indeterminate")
                    setProgress(progressMax, progressCurrent, progressIndeterminate)
                }
            }

            return builder.build()
        }
    }
}
